### ISSUE TITLE
epics.ca.ChannelAccessException: loading Epics CA DLL failed: Could not find module Com.dll in windows10

### ISSUE THREAD

**Issue by Tinanuaa** [pyepics/pyepics#217]
I installed pyepics, bluesky, ophyd and databroker in my windows 10 system, just following the steps in bluesky tutorial (https://nsls-ii.github.io/bluesky/tutorial.html#tutorial-capture-data), my python version is 3.8.7. And whenever I want to initialise a RunEngine with RE = RunEngine({}), it failes with an error showing "epics.ca.ChannelAccessException: loading Epics CA DLL failed: Could not find module Com.dll". But I can manually find the Com.dll in my virtual environment "env38\Lib\site-packages\epics\clibs\win64\Com.dll". 

The package versions I am using is 
bluesky==1.6.7
bluesky-live==0.0.7
databroker==1.2.0
ophyd==1.6.0
pyepics==3.4.3

The exception is showing as below:
![image](https://user-images.githubusercontent.com/877923/106976781-1ee10a80-67ad-11eb-862e-197acf89bb13.png)

Any help would be appreciated.

**Comment by newville** [comment]
@Tinanuaa Hm, I'm not entirely sure what's wrong.   What does

    >>> import epics
    >>> epics.ca.find_libca()
    >>> epics.ca.initialize_libca()

report?   From your output I believe these should both point to "C:\Users\...\site-packages\epics\clibs\win64\ca.dll", and that file should exist.  Can you confirm that?

Is it possible that some other library has loaded some other "ca.dll"?

**Comment by Tinanuaa** [comment]
Hi @newville , Thanks for the quick reply, the find_libca works fine with the right path but the initialize_libca just failed with the same error, not able to find Com.dll
![image](https://user-images.githubusercontent.com/877923/106979816-328f6f80-67b3-11eb-93b4-2f5e6e570d97.png)
And I've debugged into initialize_libca to see what it does was loading the dll with ctypes.windll.LoadLibrary(r"C:\Users\fengt\Desktop\projects\bluesky_tutorials\env\Lib\site-packages\epics\clibs\win64\ca.dll"), right? 
So I just run the ctypes.windll.LoadLibrary(r"C:\Users\fengt\Desktop\projects\bluesky_tutorials\env\Lib\site-packages\epics\clibs\win64\ca.dll") directly, and it failed as well, the same error.

**Comment by newville** [comment]
@Tinanuaa Hm, that seems weird that it can find `ca.dll` but not `Com.dll` in the same directory. 

Can you verify that `ca.dll` and `Com.dll` are actually in the folder   "C:\Users...\site-packages\epics\clibs\win64\", and are not zero-size?  

I might suggest re-installing pyepics with  `pip install --force-reinstall pyepics`.

Just to be clear, I run several different apps to run my beamline on Windows 10 (mostly using Py38), so I know that it should work.   I just tried creating a fresh Python 3.8 environment, activating it, and `pip installing pyepics`, and it all worked fine. 

Note that the `pyepics` package on conda-forge is severely misconfigured,  labeling unnecessary packages as dependencies (absurdly prescribing that pyepics depends on Perl), and silently and automatically mis-sets the environmental variable `PYEPICS_LIBCA` to point to a dll other than the one we provide.   

I cannot explain that package in any terms other than "intentionally broken by design".  I have nothing to do with this version.   More than "not recommending" it, I strongly recommend actively avoiding that.   Honestly, the very existence of that deliberately misconfigured packaging of my work has soured me permanently on the epics community and all people involved with that effort.   You'll note the decided lack of effort on this project in past years that strongly correlates with that.

**Comment by Tinanuaa** [comment]
@newville Yes, they are in the folder, as shown below, and not zero size
![image](https://user-images.githubusercontent.com/877923/106991848-37f8b400-67cb-11eb-85d5-c4c69148e3f6.png)

I installed python only, not from conda or miniconda, and tried with python3.6 python 3.8, python3.9 to create virtual environment and tried in different directories, but got no luck.  By the way, do I need to set the PYEPICS_LIBCA environment variable?

**Comment by newville** [comment]
@Tinanuaa  Hm, the sizes of the ca.dll and Com.dll look right to me.  That seems like it should work.  

You should not need to set PYEPICS_LIBCA unless you want to point to a different version of ca.dll/Com.dll.   But if `epics.ca.find_libca()` finds the one in `epics/clibs/win64`, then PYEPICS_LIBCA is probably not set -- that is the one it would find by default.   FWIW, I basically never set  PYEPICS_LIBCA.  But I also never, ever install the `pyepics` package from conda-forge and have never tried anything associated with bluesky.

Some other things to try (each from a fresh Python session):

a)  what does 

     C:\> python
     >>> import os, epics
     >>> print(os.environ.get('PYEPICS_LIBCA'))
     >>> epics.ca.find_libCom()
     >>> epics.ca.find_libca()

print out?

b)  If 'PYEPICS_LIBCA' is not empty / None, try

     C:\> python
     >>> import os, epics
     >>> os.environ.pop('PYEPICS_LIBCA')
     >>> epics.ca.find_libCom()
     >>> epics.ca.find_libca()
     >>> epics.ca.initialize_libca()


c) If still not working, try editing site-packages/epics.ca.py, around line 370 to 380, and just comment out the pre-loading of libCom on Windows -- I'm pretty sure it is not needed and am not entirely sure why it is there (I suspect a long-unneeded legacy based on incorrect assumptions) to read:

```python
    if os.name == 'nt':
        load_dll = ctypes.windll.LoadLibrary
    else:
        load_dll = ctypes.cdll.LoadLibrary
    try:
        # force loading the chosen version of libCom
        #COMMENT OUT# if os.name == 'nt':
        #COMMENT OUT#   load_dll(find_libCom())
        libca = load_dll(find_libca())
    except Exception as exc:
        raise ChannelAccessException('loading Epics CA DLL failed: ' + str(exc))
```

and then try the same tests.  

Sorry for all the trouble!

**Comment by hhslepicka** [comment]
# TL;DR;
```
Installing pyepics with Python from conda-forge works fine... all instructions are below. Skip to the end...
```


# About pyepics-feedstock at conda-forge

Let's go by parts, as Jack the Ripper said...


>  pyepics package on conda-forge is severely misconfigured, labeling unnecessary packages as dependencies (absurdly prescribing that pyepics depends on Perl), and silently and automatically mis-sets the environmental variable PYEPICS_LIBCA to point to a dll other than the one we provide.


pyepics feedstock on conda-forge is not severely misconfigured or linking to unnecessary packages as dependencies. Here is the evidence of that: https://github.com/conda-forge/pyepics-feedstock/blob/master/recipe/meta.yaml#L21-L26. The dependency on Perl that you mention comes from the epics-base that is listed as dependency of the pyepics feedstock.
There were already many past discussions about that on this forum here and the issues with using the provided libraries with the package.
That is the reason why the pyepics package at conda-forge depends on epics-base.

> I cannot explain that package in any terms other than "intentionally broken by design".

Not really.

> Honestly, the very existence of that deliberately misconfigured packaging of my work has soured me permanently on the epics community and all people involved with that effort. 

I'm sorry about that and if in any way I am part of this. I respect your work on this and many other projects. If you think that the package is misconfigured in any way, you are more than welcome to join the maintainers list of the pyepics-feedstock or open an issue or PR there suggesting the appropriate fixes. (https://github.com/conda-forge/pyepics-feedstock)

>You'll note the decided lack of effort on this project in past years that strongly correlates with that.

This is an excellent package and used by many people. I'm sorry that having pyepics available via conda-forge caused all this trouble.



# Conda install of pyepics and Windows 10.

Here is the procedure executed starting from a brand new installation of conda on my Windows 10 VM.

```
(base) C:\Users\Hugo>conda create -n pyepics-test python=3.8 pyepics -c conda-forge
Collecting package metadata (current_repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.9.2

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: C:\Users\Hugo\miniconda3\envs\pyepics-test

  added / updated specs:
    - pyepics
    - python=3.8


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2020.12.5          |   py38haa244fe_1         144 KB  conda-forge
    epics-base-7.0.4           |       ha925a31_0         2.8 MB  conda-forge
    libblas-3.9.0              |            8_mkl         3.9 MB  conda-forge
    libcblas-3.9.0             |            8_mkl         3.9 MB  conda-forge
    liblapack-3.9.0            |            8_mkl         3.9 MB  conda-forge
    numpy-1.20.0               |   py38h0cc643e_0         5.3 MB  conda-forge
    perl-5.32.0.1              |       h57928b3_0        28.1 MB  conda-forge
    pip-21.0.1                 |     pyhd8ed1ab_0         1.1 MB  conda-forge
    pyepics-3.4.3              |   py38haa244fe_1         185 KB  conda-forge
    python-3.8.6               |h7840368_5_cpython        19.1 MB  conda-forge
    python_abi-3.8             |           1_cp38           4 KB  conda-forge
    setuptools-49.6.0          |   py38haa244fe_3         962 KB  conda-forge
    wincertstore-0.2           |py38haa244fe_1006          15 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        69.3 MB

The following NEW packages will be INSTALLED:

  ca-certificates    conda-forge/win-64::ca-certificates-2020.12.5-h5b45459_0
  certifi            conda-forge/win-64::certifi-2020.12.5-py38haa244fe_1
  epics-base         conda-forge/win-64::epics-base-7.0.4-ha925a31_0
  intel-openmp       conda-forge/win-64::intel-openmp-2020.3-h57928b3_311
  libblas            conda-forge/win-64::libblas-3.9.0-8_mkl
  libcblas           conda-forge/win-64::libcblas-3.9.0-8_mkl
  liblapack          conda-forge/win-64::liblapack-3.9.0-8_mkl
  mkl                conda-forge/win-64::mkl-2020.4-hb70f87d_311
  numpy              conda-forge/win-64::numpy-1.20.0-py38h0cc643e_0
  openssl            conda-forge/win-64::openssl-1.1.1i-h8ffe710_0
  perl               conda-forge/win-64::perl-5.32.0.1-h57928b3_0
  pip                conda-forge/noarch::pip-21.0.1-pyhd8ed1ab_0
  pyepics            conda-forge/win-64::pyepics-3.4.3-py38haa244fe_1
  pyparsing          conda-forge/noarch::pyparsing-2.4.7-pyh9f0ad1d_0
  python             conda-forge/win-64::python-3.8.6-h7840368_5_cpython
  python_abi         conda-forge/win-64::python_abi-3.8-1_cp38
  setuptools         conda-forge/win-64::setuptools-49.6.0-py38haa244fe_3
  sqlite             conda-forge/win-64::sqlite-3.34.0-h8ffe710_0
  vc                 conda-forge/win-64::vc-14.2-hb210afc_2
  vs2015_runtime     conda-forge/win-64::vs2015_runtime-14.28.29325-h5e1d092_0
  wheel              conda-forge/noarch::wheel-0.36.2-pyhd3deb0d_0
  wincertstore       conda-forge/win-64::wincertstore-0.2-py38haa244fe_1006


Proceed ([y]/n)? y



Downloading and Extracting Packages
python_abi-3.8       | 4 KB      | ############################################################################ | 100%
libblas-3.9.0        | 3.9 MB    | ############################################################################ | 100%
pip-21.0.1           | 1.1 MB    | ############################################################################ | 100%
epics-base-7.0.4     | 2.8 MB    | ############################################################################ | 100%
libcblas-3.9.0       | 3.9 MB    | ############################################################################ | 100%
setuptools-49.6.0    | 962 KB    | ############################################################################ | 100%
perl-5.32.0.1        | 28.1 MB   | ############################################################################ | 100%
pyepics-3.4.3        | 185 KB    | ############################################################################ | 100%
wincertstore-0.2     | 15 KB     | ############################################################################ | 100%
python-3.8.6         | 19.1 MB   | ############################################################################ | 100%
certifi-2020.12.5    | 144 KB    | ############################################################################ | 100%
liblapack-3.9.0      | 3.9 MB    | ############################################################################ | 100%
numpy-1.20.0         | 5.3 MB    | ############################################################################ | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate pyepics-test
#
# To deactivate an active environment, use
#
#     $ conda deactivate



(base) C:\Users\Hugo>conda activate pyepics-test

(pyepics-test) C:\Users\Hugo>set
ALLUSERSPROFILE=C:\ProgramData
APPDATA=C:\Users\Hugo\AppData\Roaming
CommonProgramFiles=C:\Program Files\Common Files
CommonProgramFiles(x86)=C:\Program Files (x86)\Common Files
CommonProgramW6432=C:\Program Files\Common Files
COMPUTERNAME=DESKTOP-GD4IOL8
ComSpec=C:\WINDOWS\system32\cmd.exe
CONDA_DEFAULT_ENV=pyepics-test
CONDA_EXE=C:\Users\Hugo\miniconda3\Scripts\conda.exe
CONDA_PREFIX=C:\Users\Hugo\miniconda3\envs\pyepics-test
CONDA_PREFIX_1=C:\Users\Hugo\miniconda3
CONDA_PROMPT_MODIFIER=(pyepics-test)
CONDA_PYTHON_EXE=C:\Users\Hugo\miniconda3\python.exe
CONDA_SHLVL=2
DriverData=C:\Windows\System32\Drivers\DriverData
EPICS_BASE=C:\Users\Hugo\miniconda3\envs\pyepics-test\epics
EPICS_BASE_HOST_BIN=C:\Users\Hugo\miniconda3\envs\pyepics-test\epics\bin\windows-x64
EPICS_BASE_VERSION=7.0.4
EPICS_HOST_ARCH=windows-x64
FPS_BROWSER_APP_PROFILE_STRING=Internet Explorer
FPS_BROWSER_USER_PROFILE_STRING=Default
HOMEDRIVE=C:
HOMEPATH=\Users\Hugo
LIB=C:\Users\Hugo\miniconda3\envs\fortran-test\Library\lib;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
\VC\Tools\MSVC\14.16.27023\lib\x64;C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\um\x64;C:\Program Files (x86)
\Windows Kits\10\Lib\10.0.18362.0\ucrt_enclave\x64;C:\Users\Hugo\miniconda3\envs\fortran-test\Library\bin;C:\Users\Hugo\
miniconda3\envs\fortran-test\Library\mingw-w64\bin;
LOCALAPPDATA=C:\Users\Hugo\AppData\Local
LOGONSERVER=\\DESKTOP-GD4IOL8
NUMBER_OF_PROCESSORS=2
OneDrive=C:\Users\Hugo\OneDrive
OS=Windows_NT
Path=C:\Users\Hugo\miniconda3\envs\pyepics-test\epics\bin\windows-x64;C:\Users\Hugo\miniconda3\envs\pyepics-test;C:\User
s\Hugo\miniconda3\envs\pyepics-test\Library\mingw-w64\bin;C:\Users\Hugo\miniconda3\envs\pyepics-test\Library\usr\bin;C:\
Users\Hugo\miniconda3\envs\pyepics-test\Library\bin;C:\Users\Hugo\miniconda3\envs\pyepics-test\Scripts;C:\Users\Hugo\min
iconda3\envs\pyepics-test\bin;C:\Users\Hugo\miniconda3\condabin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;
C:\WINDOWS\System32\WindowsPowerShell\v1.0;C:\Git\cmd;C:\WINDOWS\System32\OpenSSH;C:\Users\Hugo\miniconda3\envs\fortran-
test\Library\mingw-w64\bin;C:\Users\Hugo\miniconda3\envs\fortran-test\Library\bin;C:\Users\Hugo\miniconda3\envs\fortran-
test\Library\lib;C:\Program Files (x86)\Windows Kits\10\Lib\10.0.18362.0\ucrt_enclave\x64;C:\Program Files (x86)\Windows
 Kits\10\Lib\10.0.18362.0\um\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.16.27023
\lib\x64;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Git\cmd;C
:\Us;C:\Users\Hugo\AppData\Local\Microsoft\WindowsApps;.
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
PROCESSOR_ARCHITECTURE=AMD64
PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
PROCESSOR_LEVEL=6
PROCESSOR_REVISION=9e0d
ProgramData=C:\ProgramData
ProgramFiles=C:\Program Files
ProgramFiles(x86)=C:\Program Files (x86)
ProgramW6432=C:\Program Files
PROMPT=(pyepics-test) $P$G
PSModulePath=C:\Program Files\WindowsPowerShell\Modules;C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules
PUBLIC=C:\Users\Public
PYEPICS_LIBCA=C:\Users\Hugo\miniconda3\envs\pyepics-test\epics\bin\windows-x64\ca.dll
RPN_DEFNS=C:\WINDOWS\defns.rpn
SESSIONNAME=Console
SystemDrive=C:
SystemRoot=C:\WINDOWS
TEMP=C:\Users\Hugo\AppData\Local\Temp
TMP=C:\Users\Hugo\AppData\Local\Temp
USERDOMAIN=DESKTOP-GD4IOL8
USERDOMAIN_ROAMINGPROFILE=DESKTOP-GD4IOL8
USERNAME=Hugo
USERPROFILE=C:\Users\Hugo
VS140COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\
windir=C:\WINDOWS

(pyepics-test) C:\Users\Hugo>



(pyepics-test) C:\Users\Hugo>python
Python 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 22:54:47) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import epics
>>> epics.ca.find_libca()
'C:\\Users\\Hugo\\miniconda3\\envs\\pyepics-test\\epics\\bin\\windows-x64\\ca.dll'
>>> epics.ca.initialize_libca()
<WinDLL 'C:\Users\Hugo\miniconda3\envs\pyepics-test\epics\bin\windows-x64\ca.dll', handle 7ffeb1450000 at 0x263fcb57880>

>>>
```

**Comment by tacaswell** [comment]
@Tinanuaa The issue is likely that you are missing the VC++ runtime dll, try:

```
pip install msvc-runtime
```

or installing visual studio (my guess is that this or some other incidental installation is why it works on Matt's machines).

The conda-forge packaging makes sure that detail like this are handled correctly.  Matt's unprofessional vitriol aside, the conda-forge packaging is a good-faith effort to correctly package pyepics in larger ecosystem.

xref https://github.com/matplotlib/matplotlib/issues/14558#issuecomment-502769057 to explanation of this issue from someone I consider authoritative about Python and compiled extensions on windows.

**Comment by newville** [comment]
>> pyepics package on conda-forge is severely misconfigured, labeling unnecessary packages as dependencies (absurdly prescribing that pyepics depends on Perl), and silently and automatically mis-sets the environmental variable PYEPICS_LIBCA to point to a dll other than the one we provide.

> pyepics feedstock on conda-forge is not severely misconfigured or linking to unnecessary packages as dependencies. Here is the evidence of that: https://github.com/conda-forge/pyepics-feedstock/blob/master/recipe/meta.yaml#L21-L26. The dependency on Perl that you mention comes from the epics-base that is listed as dependency of the pyepics feedstock.

pyepics does not depend on epics base.  You can claim that it does, but it continues to not be true.   Pyepics depends on libca and libCom, which are provided. This runs and works for a great number of people on a great number of machines and installations without epics base installed.   

The conda package claims that pyepics depends on epics base.  This is wrong.  The people who write the conda package know that this is wrong and they continue to say that it is not.

> There were already many past discussions about that on this forum here and the issues with using the provided libraries with the package.
> That is the reason why the pyepics package at conda-forge depends on epics-base.

Citation needed.

**Comment by hhslepicka** [comment]
Where does libca and libCom comes from?

**Comment by newville** [comment]
@hhslepicka 

> Where does libca and libCom comes from?

they are packaged with pyepics and included in the installation, see https://github.com/pyepics/pyepics/tree/master/epics/clibs
The people making the conda package must know this.    What I am saying is that the conda package should not depend on another conda package called "epics-base" -- this is not really a dependency.  You can verify that by doing `pip install pyepics`. 

These specially-built libraries have been included for Windows for basically the entirety of pyepics development. Like, for a decade.  For MacOs and Linux, they've been included for somewhat less time.   These work for a great many people on many platforms and machines - we run the Windows version of this on dozens of machines at our beamlines without any special effort or other packages installed.   I can certainly believe that they could have flaws, though I don't precisely understand the problem the user here is having.

**Comment by hhslepicka** [comment]
@newville I will answer my own question above... `Where does libca and libCom comes from?` - They come from epics base. Their source code is there and they are compiled as part of epics base.  The fact that you compile and decide to copy/keep binaries from epics base with pyepics is a design decision that you picked and I have nothing against it.

For conda-forge we do depend on epics-base due to a couple of reasons:
- One recipe must not vendor binaries from another or from a third-party package.
- Many users of pyepics often use the Channel Access tools (caget, caput, camonitor)
- Having EPICS Base as a dependency allows us to have pyepics as a `noarch` or architecture independent package as it relies on the architecture specific package `epics-base`.

If you use pyepics from conda-forge you don't need to solve this: https://github.com/pyepics/pyepics/issues/114
Since epics-base for aarch64 or even the newer osx-arm64 can easily supported as we can also easily bump the epics base version without the need to rely on binaries compiled with a certain compiler version or os version that you have available.

This boils down also to preference, opinion and other aspects that are personal and I respect your opinion.

The last statements that I would like to add here are:
- Please don't be sour towards people working on a way to make your library available in a very easy and alternative way to many other people. At least speaking for myself, I mean no disrespect towards you or your work.
- Don't let this keep you from evolving/working on your library that is used by many. Embrace the open-source aspect of it and if you don't like/agree with something simply deny the Pull Request.
- If someone shows up here with an issue and you suspect that it is related to conda-forge or equivalent, please feel free to tag me and I will do my best to help this person or redirect the issue to the pyepics-feedstock.


Have a great weekend! Peace!

![image](https://user-images.githubusercontent.com/8185425/107094813-8c486780-67bc-11eb-800f-a19a5fece543.png)

**Comment by Tinanuaa** [comment]
@newville Thanks for the , I tried them and got result as below:
a) 
![image](https://user-images.githubusercontent.com/877923/107144377-edc03000-698e-11eb-9c7f-96d3a94a2ce5.png)
b) 'PYEPICS_LIBCA' is None, so I skip this step
c) I got error as not able to load ca.dll after commenting out the lines
![image](https://user-images.githubusercontent.com/877923/107144559-10067d80-6990-11eb-940a-ba8aaf64d404.png)

@tacaswell I tried install msvc-runtime and Visual Studio, but both failed to solve this problem

**Comment by newville** [comment]
@Tinanuaa Hm, very sorry for the trouble. I don't think I've ever seen that or heard about other people having that.  We definitely run python with pyepics on lots of Windows 10 machines, but maybe we're assuming that some runtime dlls are installed.  

For me (Windows10, Anaconda Python, with cygwin tools installed), I see:

```
C:\>ldd C:\Users\xas_user\python\Lib\site-packages\epics\clibs\win64\ca.dll
        ntdll.dll => /cygdrive/c/Windows/SYSTEM32/ntdll.dll (0x7ff9223c0000)
        KERNEL32.DLL => /cygdrive/c/Windows/System32/KERNEL32.DLL (0x7ff920910000)
        KERNELBASE.dll => /cygdrive/c/Windows/System32/KERNELBASE.dll (0x7ff91f3e0000)
        WS2_32.dll => /cygdrive/c/Windows/System32/WS2_32.dll (0x7ff9206d0000)
        RPCRT4.dll => /cygdrive/c/Windows/System32/RPCRT4.dll (0x7ff9205b0000)
        MSVCP100.dll => /cygdrive/c/Windows/SYSTEM32/MSVCP100.dll (0x722b0000)
        MSVCR100.dll => /cygdrive/c/Windows/SYSTEM32/MSVCR100.dll (0x72410000)
        Com.dll => /cygdrive/c/Users/xas_user/python/Lib/site-packages/epics/clibs/win64/Com.dll (0x7ff8fed40000)
        ADVAPI32.dll => /cygdrive/c/Windows/System32/ADVAPI32.dll (0x7ff9222d0000)
        msvcrt.dll => /cygdrive/c/Windows/System32/msvcrt.dll (0x7ff9209f0000)
        sechost.dll => /cygdrive/c/Windows/System32/sechost.dll (0x7ff920a90000)

C:\>ldd C:\Users\xas_user\python\Lib\site-packages\epics\clibs\win64\Com.dll
        ntdll.dll => /cygdrive/c/Windows/SYSTEM32/ntdll.dll (0x7ff9223c0000)
        KERNEL32.DLL => /cygdrive/c/Windows/System32/KERNEL32.DLL (0x7ff920910000)
        KERNELBASE.dll => /cygdrive/c/Windows/System32/KERNELBASE.dll (0x7ff91f3e0000)
        WS2_32.dll => /cygdrive/c/Windows/System32/WS2_32.dll (0x7ff9206d0000)
        RPCRT4.dll => /cygdrive/c/Windows/System32/RPCRT4.dll (0x7ff9205b0000)
        ADVAPI32.dll => /cygdrive/c/Windows/System32/ADVAPI32.dll (0x7ff9222d0000)
        msvcrt.dll => /cygdrive/c/Windows/System32/msvcrt.dll (0x7ff9209f0000)
        sechost.dll => /cygdrive/c/Windows/System32/sechost.dll (0x7ff920a90000)
        MSVCR100.dll => /cygdrive/c/Windows/SYSTEM32/MSVCR100.dll (0x72410000)


C:\>ldd C:\Users\xas_user\python\Python.exe
        ntdll.dll => /cygdrive/c/Windows/SYSTEM32/ntdll.dll (0x7ff9223c0000)
        KERNEL32.DLL => /cygdrive/c/Windows/System32/KERNEL32.DLL (0x7ff920910000)
        KERNELBASE.dll => /cygdrive/c/Windows/System32/KERNELBASE.dll (0x7ff91f3e0000)
        ucrtbase.dll => /cygdrive/c/Windows/System32/ucrtbase.dll (0x7ff91f9a0000)
        python38.dll => /cygdrive/c/Users/xas_user/xraylarch/python38.dll (0x7ff8e5920000)
        VCRUNTIME140.dll => /cygdrive/c/Users/xas_user/xraylarch/VCRUNTIME140.dll (0x7ff90f0a0000)
        SHLWAPI.dll => /cygdrive/c/Windows/System32/SHLWAPI.dll (0x7ff920d10000)
        VERSION.dll => /cygdrive/c/Windows/SYSTEM32/VERSION.dll (0x7ff91abf0000)
        msvcrt.dll => /cygdrive/c/Windows/System32/msvcrt.dll (0x7ff9209f0000)
        combase.dll => /cygdrive/c/Windows/System32/combase.dll (0x7ff920e50000)
        RPCRT4.dll => /cygdrive/c/Windows/System32/RPCRT4.dll (0x7ff9205b0000)
        bcryptPrimitives.dll => /cygdrive/c/Windows/System32/bcryptPrimitives.dll (0x7ff91fb00000)
        GDI32.dll => /cygdrive/c/Windows/System32/GDI32.dll (0x7ff920740000)
        win32u.dll => /cygdrive/c/Windows/System32/win32u.dll (0x7ff91faa0000)
        gdi32full.dll => /cygdrive/c/Windows/System32/gdi32full.dll (0x7ff91f690000)
        msvcp_win.dll => /cygdrive/c/Windows/System32/msvcp_win.dll (0x7ff91fb90000)
        USER32.dll => /cygdrive/c/Windows/System32/USER32.dll (0x7ff920770000)
        WS2_32.dll => /cygdrive/c/Windows/System32/WS2_32.dll (0x7ff9206d0000)
        ADVAPI32.dll => /cygdrive/c/Windows/System32/ADVAPI32.dll (0x7ff9222d0000)
        sechost.dll => /cygdrive/c/Windows/System32/sechost.dll (0x7ff920a90000)
```

So, knowing that Python.exe runs, the  DLLs that ca.dll and Com.dll want to load that could be possibly be missing are MSVCP100.dll and MSVCR100.dll, which are part of the Microsoft runtime library - actually "the older runtime library", so it might make sense that these could be missing on some machines.

I believe these MSVC*100.dll files can be downloaded and installed from 
    [https://www.microsoft.com/en-us/download/details.aspx?id=14632]( https://www.microsoft.com/en-us/download/details.aspx?id=14632)
or by looking for the "Microsoft Visual C++ 2010 Redistributable Package (x64)".    Does installing that help?


These versions may seem outdated, but until very recently, I know that many people were still using Windows 7.   Oddly, I recently installed Anaconda Python/pyepics-from-pypi on a fresh Windows10 laptop and it worked.  Maybe all the machines I use have this older runtime installed by default (or before the admins hand it to me;)).  

Perhaps we should include these older runtime libs with ca.dll and Com.dll, or switch to updated ca.dll/Com.dll built with the newer runtime versions. I think would then stop working with older systems (but at most facilities using epics, those are probably no longer officially supported anyway).  That would seem like a good time to drop 32-bit Windows too.

**Comment by Tinanuaa** [comment]
@newville It worked!!! Thanks for all the help, and I learned a lot from this troubleshooting, so still worth the time,hahahahaha.
![image](https://user-images.githubusercontent.com/877923/107166824-3c131480-6a0b-11eb-8e7f-921e3381cf7e.png)

**Comment by newville** [comment]
@hhslepicka 
> I will answer my own question above... Where does libca and libCom comes from? - They come from epics base. 

Where these shared libraries come from for the `pyepics` package is documented in the READMEs in the `epics/clibs` folders in the pyepics source kit.  They do not come from a conda package called `epics-base`.  Perhaps you are confusing that package with Epics base. 

> Their source code is there and they are compiled as part of epics base. 

I do not understand what you are saying.  The source code for what is where?

For pyepics, we have tended to use approximately "stable release" at the time of making the shared libraries.  But we do not update the libraries very often - the current ones are 2 years old - and it may be time for an update.  As the Issue shows, on Windows, the 2010 runtime libs are needed.  The Linux libs were built with no-longer-supported CentOS 6.    

At this point, I would probably again take the versions compiled by the Epics base Jenkins server for Windows, MacOS, and others.  For Linux, I have compiled these on virtual machines - I would probably use CentOS 7 (gcc 4.8.5, glibc ~2.20 (I forget) ) now if the Jenkins server used a more modern system and glibc.

Anyway, Libca is designed to be extremely stable and maintain an API that is tolerant of clients/hosts using different generations.  From my perspective, as long as we're using base R3.15 or higher, all the features will work and the main goal is to maintain compatibility with as many current systems as possible.  Include the MSVC runtimes might be a good way to future-proof the Windows dlls.


> The fact that you compile and decide to copy/keep binaries from epics base with pyepics is a design decision that you picked and I have nothing against it.

Everything about `pyepics` exists outside of `conda` and happens to pre-date the existence of `conda`.  Again (because people may not understand this or willfully ignore it), `pyepics` can be installed with `pip install` independent of `Anaconda Python`, and does not require that any other version of Epics base be previously installed on the computer in order for `pyepics` to work.   It's been this way forever for Windows. 

> For conda-forge we do depend on epics-base due to a couple of reasons:
>
> One recipe must not vendor binaries from another or from a third-party package.

Sorry, I do not understand this.  Like, I cannot quite parse the sentence unless "vendor" is a verb, which I don't really understand.  I'm probably also lost by what "third-party" means here. You are talking about a project to package `pyepics` for `conda`, right? 

> Many users of pyepics often use the Channel Access tools (caget, caput, camonitor)

Certainly, the `epics-base` conda package is convenient for people wanting to use Epics on their computers and happen to be using Anaconda Python.  Perhaps changing the pyepics testing framework to use the `softIoc` from this package would be a good way to improve the testing.

But the command line CA tools are not *required* for using `pyepics`.

I think what you are saying here is that you are *choosing* to install the `epics-base` conda package and *choosing* to assert that the `pyepics` conda package from `conda-forge` depends on the `epics-base` conda package from `conda-forge`. 

That is exactly what I am saying:  A package for `pyepics`, whether from `pip`, `conda`, `homebrew`, `apt`, `yum/dnf`, or any other packaging system does *not* need to rely on a separate packaging of Epics base.
There are, in fact, conda packagings of `pyepics` on other channels that do not depend on a conda package for Epics base.   The `conda-forge` package for `pyepics` asserts that `pyepics` depends on a package called `epics-base`.  That is a simply not true.


> Having EPICS Base as a dependency allows us to have pyepics as a noarch or architecture independent package as it relies on the architecture specific package epics-base.

Whether they are `noarch` or use trivial `conda convert` is sort of a separate point (and, again, a specific feature only of `conda`).  In my experience, conda packaging, the ability to use `noarch`, and dependency resolution have gone through a few painful opaque developments over the past few years.

Again, `pip install` works and not using wheels built per-architecture.   The ca shared-libraries are included as "package data" and found at runtime.

> If you use pyepics from conda-forge you don't need to solve this: #114

OK.  But then you need to use conda-forge....

> Since epics-base for aarch64 or even the newer osx-arm64 can easily supported as we can also easily bump the epics base version without the need to rely on binaries compiled with a certain compiler version or os version that you have available.

Well, those could be pushed to `pyepics`, right?  And then they could be independent of Anaconda Python.  Contributions welcome.

> This boils down also to preference, opinion and other aspects that are personal and I respect your opinion.

Well, except that the conda-forge package for pyepics asserts dependencies that simply are not true.  Package dependencies are not really a matter of opinion.

> The last statements that I would like to add here are:
>
> Please don't be sour towards people working on a way to make your library available in a very easy and alternative 
> way to many other people. At least speaking for myself, I mean no disrespect towards you or your work.

Hm, I believe you that you do not. I am not sure I can say that about everyone.  Still, the point here is that the `conda-forge` packaging of `pyepics` is asserting things about `pyepics` that are not true.

> Don't let this keep you from evolving/working on your library that is used by many. Embrace the open-source 
> aspect of it and if you don't like/agree with something simply deny the Pull Request.

It turns out that I do not need to be sold on the benefits of open source.   Please fix the `pyepics` conda channel to not depend on the `epics-base` package and not overwrite `PYEPICS_LIBCA` on startup.   Those are choices and they should be left to the user.

> If someone shows up here with an issue and you suspect that it is related to conda-forge or equivalent, please feel free to tag me and I will do my best to help this person or redirect the issue to the pyepics-feedstock.

OK.
